### PR TITLE
feat: added varint support for metadata readonly fields

### PR DESCRIPTION
**Pull request by rajatkb**
# Description
- Adds a little endian VarINT function
- Uses the same for metadata fields 
- BigEndian machine needs to be handled

### Issue
Check : [url](https://github.com/spirit-labs/tektite/issues/270)

### CODE REVIEW

File: asl/encoding/encoding.go [hunk 1]
```diff
@@ -45,6 +48,16 @@ func AppendUint32ToBufferLE(buffer []byte, v uint32) []byte {
 	return binary.LittleEndian.AppendUint32(buffer, v)
 }
 
+func AppendUint32ToBufferLEAsVarInt(buffer []byte, v uint32) []byte {
```
Comment (purplefox) [inline asl/encoding/encoding.go:51]: I don't think there is any need for these methods. There's already a binary.AppendUvarint() method which does this, and does not do extra allocations like this method does. Here's an example of its use https://github.com/spirit-labs/tektite/blob/main/kafkaprotocol/api_versions_request.go#L64

File: asl/encoding/encoding.go [hunk 2]
```diff
@@ -82,13 +95,22 @@ func ReadUint64FromBufferLE(buffer []byte, offset int) (uint64, int) {
...
 func ReadUint32FromBufferLE(buffer []byte, offset int) (uint32, int) {
+
 	if IsLittleEndian {
 		// nolint: gosec
 		return *(*uint32)(unsafe.Pointer(&buffer[offset])), offset + 4
 	}
 	return littleEndian.Uint32(buffer[offset:]), offset + 4
 }
 
+func ReadUint32FromBufferLEVarInt(buffer []byte, offset int) (uint32, int) {
```
Comment (purplefox) [inline asl/encoding/encoding.go:106]: Again, this seems unnecessary. There's a binary.Uvarint method that can be used directly to do this. Here's an example of it being used: https://github.com/spirit-labs/tektite/blob/main/kafkaprotocol/api_versions_request.go#L21

**Review by purplefox — CHANGES_REQUESTED** [review]
Hi @rajatkb - a good start, but I think things could be simplified by using the already existing methods for appending and reading a varint (see comments).

File: asl/encoding/encoding.go (hunk 1 above)
Comment (rajatkb) [reply asl/encoding/encoding.go:51]: Thanx for the pointer. Was not aware let me check those section.

**Comment by purplefox** [comment]
Hi @rajatkb are you still working on this?

**Comment by rajatkb** [comment]
@purplefox  hey yes , I was off for last two weeks so could not come back to this. I will flesh out the PR based on comments today.

File: asl/encoding/encoding.go (hunk 1 above)
Comment (rajatkb) [reply asl/encoding/encoding.go]: Removed the helpers and using the package methods directly

File: asl/encoding/encoding.go (hunk 2 above)
Comment (rajatkb) [reply asl/encoding/encoding.go]: this is addressed now

**Comment by rajatkb** [comment]
@purplefox  can you check now. have removed the helpers

File: asl/encoding/encoding.go [hunk 3]
```diff
@@ -2,11 +2,12 @@ package encoding
```
Comment (purplefox) [inline asl/encoding/encoding.go:2]: Changes in this file seem to be unnecessary

File: asl/encoding/encoding_test.go [hunk 4]
```diff
@@ -1,9 +1,10 @@
 package encoding
```
Comment (purplefox) [inline asl/encoding/encoding_test.go:1]: changes in this file are also unnecessary

File: sst/sstable.go [hunk 5]
```diff
@@ -3,14 +3,16 @@ package sst
 import (
 	"bytes"
 	"encoding/binary"
+	"math"
```
Comment (purplefox) [inline sst/sstable.go:6]: no need to re-order imports

File: sst/sstable.go [hunk 6]
```diff
@@ -144,15 +150,17 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
 
 	// Now fill in metadata offset
 	metadataOffset := len(buff)
+
```
Comment (purplefox) [inline sst/sstable.go:153]: unnecessary new whitespace

File: sst/sstable.go [hunk 7]
```diff
@@ -161,18 +169,30 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
...
-	}, smallestKey, largestKey, minVersion, maxVersion, nil
+	}
+
+	selfTable.metadataBinary = selfTable.serializedMetdata()
+	selfTable.metadataSize = len(selfTable.metadataBinary)
+
+	return selfTable, smallestKey, largestKey, minVersion, maxVersion, nil
+}
+
+func (s *SSTable) serializedMetdata() []byte {
```
Comment (purplefox) [inline sst/sstable.go:180]: This method is not necessary and will cause extra unnecessary buffer copying. Simpler to just append the varints directly to the existing buffer.

File: sst/sstable.go [hunk 8]
```diff
@@ -161,18 +169,30 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
...
 	// To avoid copying the data buffer, we put all the meta-data at the end
-	buff := encoding.AppendUint32ToBufferLE(s.data, s.maxKeyLength)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numEntries)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numDeletes)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numPrefixDeletes)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.indexOffset)
-	buff = encoding.AppendUint64ToBufferLE(buff, s.creationTime)
-	return buff
+
+	return append(s.data, s.metadataBinary...)
```
Comment (purplefox) [inline sst/sstable.go:195]: Simpler and more efficient to just do the varint appending directly in here.

**Review by purplefox — COMMENTED** [review]
Hi @rajatkb - thanks for this. It's getting there! Things could be simplified a bit more - creating a new buffer is not necessary and makes things less efficient. Also the PR contains changes to files (e.g. import re-ordering) that aren't relevant to the PR.

File: asl/encoding/encoding.go (hunk 3 above)
Comment (rajatkb) [reply asl/encoding/encoding.go:2]: @purplefox  some of the changes are lint/formating changes. Do we have any linter enabled in the project which I can use. Vscode seems to make these changes from it standard formatting of golang

File: sst/sstable.go (hunk 8 above)
Comment (rajatkb) [reply sst/sstable.go:195]: But I will need the size of metadata buffer ahead of time during `BuildSSTable` as it's used in `SizeBytes`. Hence the need of the function `serializedMetdata` . Any suggestion around the `SizeBytes` while avoiding buffer copy ?

Comment (purplefox) [reply sst/sstable.go:195]: If you want to know the size of the metadata that you can just measure len() before and after appending it to the buffer, no?

Comment (rajatkb) [reply sst/sstable.go:195]: Yes and the new function I am about to commit does that. but that assumes the SizeByte is only called after a call to `Serialize` . Or are you suggesting to just append the metadata when creating the `SSTable struct instance`.

Comment (purplefox) [reply sst/sstable.go]: With your changes you're already creating the serialized metadata in the BuildSSTable method. But you're building that into a new buffer which you then append to the main buffer on Serialize. That seems unnecessary. Could you not just append it into the same buffer with all the rest of the data?

Comment (rajatkb) [reply sst/sstable.go]: Yes I did that.

File: asl/encoding/encoding.go (hunk 3 above)
Comment (purplefox) [reply asl/encoding/encoding.go:2]: There are no relevant changes in this file, so can just be removed from the PR

File: asl/encoding/encoding_test.go (hunk 4 above)
Comment (purplefox) [reply asl/encoding/encoding_test.go:1]: as above

File: sst/sstable.go (hunk 5 above)
Comment (purplefox) [reply sst/sstable.go:6]: regarding order of imports - I just use  goland and default settings - can you configure your IDE to not re-order imports on save of file? (I know this is possible with goland, don't know about your IDE)

**Comment by rajatkb** [comment]
@purplefox  
- cleaned up the unnecessary changes in files
- buffer copies are avoided now

File: sst/sstable.go [hunk 9]
```diff
@@ -147,32 +150,42 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
...
-	}, smallestKey, largestKey, minVersion, maxVersion, nil
+	}
+
+	buff, selfTable.metadataSize = selfTable.serializedMetdata(buff)
+	selfTable.data = buff
+
+	return selfTable, smallestKey, largestKey, minVersion, maxVersion, nil
+}
+
+func (s *SSTable) serializedMetdata(buff []byte) ([]byte, int) {
```
Comment (purplefox) [inline sst/sstable.go:175]: Looking much better. As a further simplification there's no need to have a separate method for this, just do the appends in the BuildSSTable method

Comment (rajatkb) [reply sst/sstable.go:175]: I just named the function. Can do there as well.

Comment (rajatkb) [reply sst/sstable.go]: Addressed

File: sst/sstable.go [hunk 10]
```diff
@@ -147,32 +151,38 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
...
 		numEntries:       uint32(numEntries),
 		numDeletes:       uint32(numDeletes),
 		numPrefixDeletes: uint32(numPrefixDeletes),
 		indexOffset:      uint32(indexOffset),
 		creationTime:     uint64(time.Now().UTC().UnixMilli()),
-		data:             buff,
-	}, smallestKey, largestKey, minVersion, maxVersion, nil
+	}
+
+	prevBuffLen := len(buff)
```
Comment (purplefox) [inline sst/sstable.go:170]: this is now unnecessary

File: sst/sstable.go [hunk 11]
```diff
@@ -147,32 +151,38 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
...
+	buff = binary.AppendUvarint(buff, uint64(maxKeyLength))
+	buff = binary.AppendUvarint(buff, uint64(numEntries))
+	buff = binary.AppendUvarint(buff, uint64(numDeletes))
+	buff = binary.AppendUvarint(buff, uint64(numPrefixDeletes))
+	buff = binary.AppendUvarint(buff, uint64(indexOffset))
+	buff = binary.AppendUvarint(buff, selfTable.creationTime)
+
+	selfTable.metadataSize = prevBuffLen - len(buff)
```
Comment (purplefox) [inline sst/sstable.go:178]: and this

File: sst/sstable.go [hunk 12]
```diff
@@ -45,6 +46,8 @@ type SSTable struct {
 	//  │maxKeyLength bytes                         │ 4 bytes  │
 	//  ╰───────────────────────────────────────────┴──────────╯
 	data []byte
+	// Varint encoded metadata length
+	metadataSize int
```
Comment (purplefox) [inline sst/sstable.go:50]: This is not used, so can be removed.

Comment (rajatkb) [reply sst/sstable.go]: Cleaned up the metadataSize instances. Might have been pre-emptive to think this information maybe necessary. Cleaned it up for now

**Pull request merged**